lib/tree.rb

```diff
@@ -2,61 +2,157 @@ class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
 
-   def initialize(key, val)
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
+
   def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log(n))
+  # Space Complexity: O(1)
```

Comment:
Since you are using a recursive `add_helper` you actually have O(log n) space complexity.  Remember recursion does involve the system stack.

This also applies to your other recursive methods.